x-pack/plugins/session_view/public/components/process_tree/helpers.ts

```diff
@@ -0,0 +1,170 @@
+  orphans?.forEach((process) => {
+    const parentProcess = processMap[process.getDetails().process.parent.entity_id];
+
+    if (parentProcess) {
+      process.parent = parentProcess; // handy for recursive operations (like auto expand)
+
+      parentProcess.children.push(process);
+    } else {
+      newOrphans.push(process);
+    }
+  });
+
+  return newOrphans;
+};
+
+// given a plain text searchQuery, iterates over all processes in processMap
+// and marks ones which match the below text (currently what is rendered in the process line item)
+// process.searchMatched is used by process_tree_node to highlight the text which matched the search
+// this funtion also returns a list of process results which is used by session_view_search_bar to drive
+// result navigation UX
+// FYI: this function mutates properties of models contained in processMap
+export const searchProcessTree = (processMap: ProcessMap, searchQuery: string | undefined) => {
+  const results = [];
+
+  for (const processId of Object.keys(processMap)) {
+    const process = processMap[processId];
+
+    if (searchQuery) {
+      const event = process.getDetails();
+      const { working_directory: workingDirectory, args } = event.process;
+
+      // TODO: the text we search is the same as what we render.
+      // in future we may support KQL searches to match against any property
+      // for now plain text search is limited to searching process.working_directory + process.args
+      const text = `${workingDirectory} ${args?.join(' ')}`;
+
+      process.searchMatched = text.includes(searchQuery) ? searchQuery : null;
+
+      if (process.searchMatched) {
+        results.push(process);
+      }
+    } else {
+      process.clearSearch();
+    }
+  }
+
+  return results;
+};
+
+// Iterate over all processes in processMap, and mark each process (and it's ancestors) for auto expansion if:
+// a) the process was "user entered" (aka an interactive group leader)
+// b) matches the plain text search above
+// Returns the processMap with it's processes autoExpand bool set to true or false
+// process.autoExpand is read by process_tree_node to determine whether to auto expand it's child processes.
+export const autoExpandProcessTree = (processMap: ProcessMap) => {
+  for (const processId of Object.keys(processMap)) {
+    const process = processMap[processId];
+
+    if (process.searchMatched || process.isUserEntered()) {
+      let { parent } = process;
+      const parentIdSet = new Set<string>();
+
+      while (parent && !parentIdSet.has(parent.id)) {
+        parentIdSet.add(parent.id);
+        parent.autoExpand = true;
+        parent = parent.parent;
+      }
+    }
+  }
+
+  return processMap;
+};
+
+export const processNewEvents = (
+  eventsProcessMap: ProcessMap,
+  events: ProcessEvent[] | undefined,
+  orphans: Process[],
+  sessionEntityId: string,
+  backwardDirection: boolean = false
+) => {
+  if (!events || events.length === 0) {
+    return eventsProcessMap;
```

Comment:
having this function return just the map here and then a tuple below of the map and new orphans seems strange, should it be `return [eventsProcessMap, []];` instead?

Reply:
seems to not work well with the cast here https://github.com/elastic/kibana/pull/124575/files#diff-7ac97f818929bd772776985fafd79a2d94b43362d9c7286ec9dc37164caa6b13R228